### PR TITLE
Add impersonate user metrics

### PR DESCRIPTION
**Pull request by mtakeda**
#### What problem is this solving?

Send events to Analytics Redshift when a user impersonates another user.
Currently, there are two functions used to impersonate and each one will send one different metric (with similar fields) in order to keep tracking about how they are used - there is a comment in the code telling to deprecate one of them, but both are still used)

#### How to test it?
 On [organization manager page](https://takeda--b2bstore005.myvtex.com/account#/organization) select a user to impersonate. After impersonate, click to stop impersonate. Events should appear in Analytics in one hour or so, into "vtex"."schemaless"."b2b_suite_buyerorg_data_raw" schema.

[Workspace](https://takeda--b2bstore005.myvtex.com/account#/organization)

#### Screenshots or example usage:

image

image

Types:
```
type ImpersonateUserMetric = {
  kind: 'impersonate-user-graphql-event'
  description: 'Impersonate User Action - Graphql'
}

type ImpersonateB2BUserMetric = {
  kind: 'impersonate-b2b-user-graphql-event'
  description: 'Impersonate B2B User Action - Graphql'
}

export type Metric = {
  name: 'b2b-suite-buyerorg-data'
  account: string
} & (ImpersonateUserMetric | ImpersonateB2BUserMetric)


type ImpersonatePerson = {
  email: string
  buy_org_id: string
  cost_center_id: string
  person_id: string
}

type ImpersonateFieldsMetric = {
  user?: ImpersonatePerson
  target: ImpersonatePerson
  date: string
}

type ImpersonateMetric = Metric & { fields: ImpersonateFieldsMetric }
```

Example of queries and results:

```
SELECT
    payload.account,
    payload.kind,
    payload.fields.date,
    payload.fields.target.email as target_email,
    payload.fields.target.buy_org_id as target_buy_org,
    payload.fields.target.cost_center_id as target_cc

FROM "vtex"."schemaless"."b2b_suite_buyerorg_data_raw"
WHERE payload.name = 'b2b-suite-buyerorg-data'
and payload.kind in ('impersonate-user-graphql-event')
order by ingestion_time desc
```

image

```
SELECT
    payload.account,
    payload.kind,
    payload.fields.date,
    payload.fields.user.email as user_email,
    payload.fields.user.buy_org_id as user_buy_org,
    payload.fields.user.cost_center_id as user_cc,
    payload.fields.target.email as target_email,
    payload.fields.target.buy_org_id as target_buy_org,
    payload.fields.target.cost_center_id as target_cc

FROM "vtex"."schemaless"."b2b_suite_buyerorg_data_raw"
WHERE payload.name = 'b2b-suite-buyerorg-data'
and payload.kind in ('impersonate-b2b-user-graphql-event')
order by ingestion_time desc
```
image

### CODE REVIEW

File: node/resolvers/Mutations/Users.ts
Comment (mairatma) [inline node/resolvers/Mutations/Users.ts:659]: The name of the field that this `impersonateUser` receives is `userId`, not `id`, right? Can you fix and make sure that you test this flow after the change so we're sure it's not introducing any problems?

File: node/resolvers/Mutations/Users.ts
Comment (mairatma) [inline node/resolvers/Mutations/Users.ts:666]: I believe we don't want to send the metrics if the call to impersonate fails, right? Or, if we do want to send metrics in error cases, we need to fix the `sendImpersonateB2BUserMetric` call further above to stop ignoring errors too.

File: node/resolvers/Mutations/Users.ts
Comment (mairatma) [inline node/resolvers/Mutations/Users.ts:683]: This shouldn't be awaited, right? It should run on the background.

**Review by mairatma — COMMENTED** [review]
It seems like there are two separate flows for impersonation here. Which of these was tested via the instructions you described for this PR? We should make sure to test both use cases, so can you add the instructions for testing both, and make sure they're still working as before?

**Comment by mtakeda** [comment]
> It seems like there are two separate flows for impersonation here. Which of these was tested via the instructions you described for this PR? We should make sure to test both use cases, so can you add the instructions for testing both, and make sure they're still working as before?


There is one function/code tree for impersonation and one for stopping impersonation. The instructions for testing already cover both (starting / stopping impersonation). 
I'm not sure why there are those two functions doing almost the same thing. There is a [comment](https://github.com/vtex-apps/b2b-organizations-graphql/blob/master/graphql/schema.graphql#L211) saying one of them is deprecated, but I was not confident to remove/replace it at all.

File: node/resolvers/Mutations/Users.ts
Comment (mtakeda) [reply node/resolvers/Mutations/Users.ts]: Good catch!
I've changed and tested it again. I've noticed that it was working because, in the stopping impersonate flow, it sends null anyway :man_shrugging:

File: node/resolvers/Mutations/Users.ts
Comment (mtakeda) [reply node/resolvers/Mutations/Users.ts:687]: Right again.
Fix to not send metrics to analytics on impersonate errors

**Comment by mairatma** [comment]
> I'm not sure why there are those two functions doing almost the same thing. There is a [comment](https://github.com/vtex-apps/b2b-organizations-graphql/blob/master/graphql/schema.graphql#L211) saying one of them is deprecated, but I was not confident to remove/replace it at all.

Can you try calling the deprecated graphql query directly, to be sure it doesn't fail? Deprecated queries still work, and it might still be used, like you said. The metrics will tell us later, but it's important to run this changed code at least once before merging it. If the UI doesn't call it anymore, it needs to be called directly for testing.

File: node/resolvers/Mutations/Users.ts
Comment (mtakeda) [reply node/resolvers/Mutations/Users.ts]: Right. Already fixed when looking another comment

**Comment by mtakeda** [comment]
> > I'm not sure why there are those two functions doing almost the same thing. There is a [comment](https://github.com/vtex-apps/b2b-organizations-graphql/blob/master/graphql/schema.graphql#L211) saying one of them is deprecated, but I was not confident to remove/replace it at all.
> 
> Can you try calling the deprecated graphql query directly, to be sure it doesn't fail? Deprecated queries still work, and it might still be used, like you said. The metrics will tell us later, but it's important to run this changed code at least once before merging it. If the UI doesn't call it anymore, it needs to be called directly for testing.


Actually, frontend calls both, but it could call just one in both scenarios (instead of calling a different one for each scenario). I believe someone marked one of them as deprecated but did not remove it exactly because the front is still calling them.

File: node/resolvers/Mutations/Users.ts
Comment (mairatma) [reply node/resolvers/Mutations/Users.ts]: Crazyness 🤯

**Review by mairatma — APPROVED** [review]
Lgtm! I'm just not 100% sure if you managed to run the code for both metric types yet. If so, it's good to go. Otherwise, just make sure to run both so we can be sure they're running as expected.

**Comment by mtakeda** [comment]
> Lgtm! I'm just not 100% sure if you managed to run the code for both metric types yet. If so, it's good to go. Otherwise, just make sure to run both so we can be sure they're running as expected.

Yes, I did. 
These are sample results showing both events in Analytics schema

image

**Review by mairatma — APPROVED** [review]

**Pull request merged**
